tables/Tables/TableUtil.cc

```diff
@@ -0,0 +1,321 @@
+      Int inx = parent.keywordSet().fieldNumber(subtableName);
+      if (inx >= 0) {
+        if (parent.keywordSet().type(inx) == TpTable) {
+          parent.reopenRW();
+          parent.rwKeywordSet().removeField (subtableName);
+        }
+      }
+    }
+
+    //# The logic is similar to that in Table::open.
+    rownr_t getLayout (TableDesc& desc, const String& tableName)
+    {
+      rownr_t nrow;
+      uInt format;
+      String tp;
+      AipsIO ios (Table::fileName(getFullName(tableName)));
+      uInt version = ios.getstart ("Table");
+      if (version > 3) {
+        throw TableError ("Table version " + String::toString(version) +
+                          " not supported by TableUtil in this version of Casacore");
+      }
+      if (version > 2) {
+        ios >> nrow;
+      } else {
+        uInt n;
+        ios >> n;
+        nrow = n;
+      }
+      ios >> format;
+      ios >> tp;
+      if (tp == "PlainTable") {
+	PlainTable::getLayout (desc, ios);
+      } else if (tp == "RefTable") {
+        RefTable::getLayout (desc, ios);
+      } else if (tp == "ConcatTable") {
+        ConcatTable::getLayout (desc, ios);
+      } else {
+        throw (TableInternalError
+               ("TableUtil::getLayout: unknown table kind " + tp));
+      }
+      ios.close();
+      return nrow;
+    }
+
+    TableInfo tableInfo (const String& tableName)
+      {
+        return BaseTable::tableInfo (getFullName(tableName));
+      }
+
+    String getFullName (const String& tableName)
+    {
+      // See if a subtable is given using ::.
+      String tabName;
+      std::pair<Table,String> t = tryFindSubTable (tableName, False);
+      if (t.first.isNull()  ||  t.second.empty()) {
+        tabName = Path(tableName).absoluteName();
+      } else {
+        tabName = t.first.keywordSet().tableAttributes(t.second).name();
+      }
+      return tabName;
+    }
+    
+    // Return Table and name of last part.
+    std::pair<Table,String> tryFindSubTable (const String& fullName, Bool mustExist)
+    {
+      Table tab;
+      String lastPart, msg;
+      // Split the name on :: to get the main and subtable names.
+      Vector<String> names = stringToVector(fullName, std::regex("::"));
+      AlwaysAssert (!names.empty(), AipsError);
+      // Check that no empty parts are given.
+      if (anyEQ (names, String())) {
+        msg = "empty name part given";
+      } else if (names.size() > 1  ||  mustExist) {
+        // No need to check if new table and no subtable given.
+        // Check if main table exists.
+        if (! Table::isReadable (names[0])) {
+          msg = "main table " + names[0] + " does not exist";
+        } else {
+          tab = Table(names[0]);
+          // Get name of last subtable.
+          if (names.size() > 1) {
+            lastPart = names[names.size()-1];
+            // Check if all subtables exist, maybe except last one.
+            uInt nrname = mustExist ? names.size() : names.size()-1;
```

Comment:
I'm somewhat confused what the return value of `tryFindSubTable` is in different cases: The value of `tab` may vary:
- If `mustExist` is true, `tab` is always the last table, with name `lastPart`.
- If `mustExist` is false, `tab` is the last-but-one table, and `lastPart` is the name of a subtable of `tab` that may or may not exist.

Maybe it's simpler if `tab` is always the last-but-one table: The semantics of the return value are then always the same. At the end of the function, you can then add `if (mustExist && !tab.keywordSet().isDefined(lastPart)) { msg = ...; }`

Also, please clearly document the return value in TableUtil.h.

Reply:
Changed name to findParentTable and removed Bool argument.
openTable now does the last open, which also solves problem that LockOption was not used.